### PR TITLE
Support Apple Diff timestamps

### PR DESCRIPTION
**Pull request by freeformz**
Apple's diff (at least the one shipping with macOS Ventura) does not include the fractional seconds or Timezone in its output.

This adds support for that so diffutils does not need to be installed.

I first encountered this problem using golangci-lint on macOS after upgrading to macOS Ventura.

### CODE REVIEW

File: diff/parse.go [hunk 1]
```diff
@@ -247,22 +250,21 @@ func (r *FileDiffReader) HunksReader() *HunksReader {
 // start with "---" and "+++" with the orig/new file names and
 // timestamps). Or which starts with "Only in " with dir path and filename.
 // "Only in" message is supported in POSIX locale: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/diff.html#tag_20_34_10
-func (r *FileDiffReader) ReadFileHeaders() (origName, newName string, origTimestamp, newTimestamp *time.Time, err error) {
+func (r *FileDiffReader) ReadFileHeaders() (origName, newName string, origTimestamp, newTimestamp *time.Time, origTZ, newTZ bool, err error) {
```
Comment (keegancsmith) [inline diff/parse.go:253]: this is a backwards incompatible change to the API, which will break users of this library.

The reason you are tracking this info is so you can re-marshal the diff byte for byte identically? Is that important? I would consider when we marshal back out we just marshal in UTC with tz.

File: diff/parse.go [hunk 2]
```diff
@@ -306,16 +308,25 @@ func (r *FileDiffReader) readOneFileHeader(prefix []byte) (filename string, time
...
+		ts, err = time.Parse(diffTimeParseLayout, parts[1])
 		if err != nil {
-			return "", nil, err
+			var err1 error
+			ts, err1 = time.Parse(diffTimeParseWithoutTZLayout, parts[1])
+			if err1 != nil {
+				return "", nil, hadTZ, err
+			}
+			hadTZ = false
+			ts = ts.In(time.Now().Location())
```
[the remark this answers is not in the document]
Comment (freeformz) [reply diff/parse.go:323]: The diff being generated by apple's diff program (assuming they're run locally) will be in the local time zone. I can double check that tomorrow though.

File: diff/parse.go (hunk 1 above)
Comment (freeformz) [reply diff/parse.go:253]: I wasn't sure if it was important or not. I'll make that change and adjust the tests tomorrow.

File: diff/diff.go [hunk 3]
```diff
@@ -15,15 +15,19 @@ type FileDiff struct {
...
 	// the original timestamp (nil if not present)
-	OrigTime *time.Time
+	OrigTime      *time.Time
+	OrigTimeHasTZ bool
 	// the new name of the file (often same as OrigName)
 	NewName string
 	// the new timestamp (nil if not present)
-	NewTime *time.Time
+	NewTime      *time.Time
+	NewTimeHasTZ bool
```
[the remark this answers is not in the document]
Comment (freeformz) [reply diff/diff.go:24]: I'm going to remove them.

File: diff/parse.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (freeformz) [reply diff/parse.go]: I decided to stick with UTC and use Parse.

**Comment by freeformz** [comment]
@keegancsmith I removed the tracking of TZ being present or not and modified the test to use an alternate file, if specified, for checking output.

**Review by keegancsmith — APPROVED** [review]
Thanks!

**Comment by rbespalovnc** [comment]
Hi, 

Thanks for fixing, is there a plan to make a new release?

**Comment by keegancsmith** [comment]
@rbespalovnc just tagged v0.7.0 https://github.com/sourcegraph/go-diff/releases/tag/v0.7.0

**Pull request merged**